### PR TITLE
paginate local data when displaying icons selector, fix #17522

### LINKED ISSUE #17522: Icons dropdown (for choosing a tabler icon) is very slow

### PR DESCRIPTION
**Pull request by orthagh**
It seems to be make wonders ^^

I wanted to add collections (with outgroup) for tabler icons, but the way icons are fetched (from the CSS file), I don't see any way to add this. But anyway, for the performance, it's already a good step in comparison of existing dropdown

### CODE REVIEW

**Comment by cconard96** [comment]
Both FontAwesome and Tabler Icons have metadata files in their libraries that could be used instead of scraping the CSS. `@fortawsome/fontawesome-free/metadata/icons.yml` (and a categories.yml) and `@tabler/icons/tags.json`. None of those files are currently moved to the `public/lib` folder.

**Comment by orthagh** [comment]
I just pushed a test (it's not working) to avoid losing it on computer switch.
Adding categories into the process complexifies a lot the pagination.

File: src/Html.php [hunk 1]
```diff
@@ -1161,6 +1161,10 @@ public static function includeHeader(
                 Html::requireJs('cable');
             }
 
+            //if (in_array('tablericons-definitions', $jslibs)) {
```
Comment (AdrienClairembault) [inline src/Html.php:1164]: Did you forget to un-comment ?

Comment (cconard96) [reply src/Html.php:1164]: This data probably shouldn't be imported like this at all. I suggest removing all of the on-demand loading stuff in this class and add an import inside the WebIconSelector.js file instead.

Comment (cconard96) [reply src/Html.php:1164]: This will also fix one of the Jest test failures.

Comment (orthagh) [reply src/Html.php:1164]: yes indeed

**Comment by cconard96** [comment]
FontAwesome support seems to be missing.

File: js/modules/Form/WebIconSelector.js [hunk 2]
```diff
@@ -75,35 +126,80 @@ export class WebIconSelector {
...
-                            text: cls
-                        };
-                        if (!icons.includes(entry)) {
-                            icons.push(entry);
-                        }
-                    }
-                }
+
+        var replacements = {
+            "Animals": __("Animals"),
```
Comment (cconard96) [inline js/modules/Form/WebIconSelector.js:131]: Translations should have a context.

**Review by cconard96 — CHANGES_REQUESTED** [review]
Last thing I think is to remove support for multiple icon sets from the `dropdownWebIcons` field macro in Twig and adjust everything that calls the macro. As far as I know, Manual Links are the only thing using this for FontAwesome icons. We can keep the existing selections for FA icons but only show the Tabler icons along with the current selection if it is FA (without the icon itself, or we can still scrape for the FA icons to be able to show it).

**Review by cconard96 — APPROVED** [review]

**Review by AdrienClairembault — COMMENTED** [review]
New dropdown seems fast, ok for me if that was the main purpose.

Not sure about the design tho, we are displaying a lot of text that is technical and can't be translated (also content is misaligned a bit):

image

I think it would be better to display only icons (no text) on a grid pattern, no ?

image

**Comment by orthagh** [comment]
> I think it would be better to display only icons (no text) on a grid pattern, no ?

Clearly yes, but as stated in #17522, we have these options:

- to use external lib for that (with the regular issues of non maintened one)
- do it ourselves
- just improve a bit the select2 implementation (speed and add categories).

For a quick pass (just improve responsivity), the latter option was the best. Post 11.0, we may explore the 2nd option.

**Comment by cconard96** [comment]
It doesn't really matter how it is displayed. Either way, you can see the icons and the search/text/tooltips are only in English. There is no getting around the English-only part. The only improvement we can still make is allow searching on the hidden tags Tabler has for each icon in the JSON. For example, the 2fa icon has tags for "password" and "login" among others. Again, English only. Having a grid may be more familiar to users, but I don't think it has any reasonable effect on usability for something that is only going to be rarely used.

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
> Not sure about the design tho, we are displaying a lot of text that is technical and can't be translated (also content is misaligned a bit):

It is almost the same as current design. I think we could merge this first step.

**Review by trasher — APPROVED** [review]

**Pull request merged**